### ISSUE TITLE
Comparison Benchmark

### ISSUE THREAD

**Issue by Xeoncross** [emersion/go-smtp#46]
Brad's original https://github.com/bradfitz/go-smtpd was forked by https://github.com/mhale/smtpd who (with help) added a lot of RFC compliance and general updates. They don't have streaming support so I forked the project again and made https://github.com/Xeoncross/smtpd. I actually ended up rewriting a lot of it to remove code better handled by the stdlib.

I went ahead and benchmarked this repo against mine after realizing @emersion had basically made all the same design decisions (using textproto instead of manual string/byte checking) and it seems you have done a better job again. 

Here you can see `BenchmarkNetSMTP` (mine) against `BenchmarkEmersionGoSMTP` (yours) with a basic 2k multipart/MIME email body read to `ioutil.Discard`. 

```
go test -bench=. --benchmem                    
goos: darwin
goarch: amd64
BenchmarkRawProcessingSequence-8   	   20000	     63182 ns/op	   34595 B/op	      68 allocs/op
BenchmarkNetSMTP-8                 	    2000	    717042 ns/op	  128394 B/op	     338 allocs/op
BenchmarkEmersionGoSMTP-8          	    3000	    517125 ns/op	   68882 B/op	     292 allocs/op
```

(Note: BenchmarkRawProcessingSequence is just a `net.Pipe()` with no actual sockets or TCP overhead for comparison.)

Thanks again for your work on all these email libraries.

**Comment by emersion** [comment]
Thanks for the benchmark! That's good to hear, but I bet performance is something that could still be improved in go-smtp. :)

**Comment by Xeoncross** [comment]
I spoke too soon. I realized I was using my streaming multipart/mime parser for my benchmark - but just reading from an io.Reader for yours. Adding https://github.com/Xeoncross/mimestream to  your go-smtp benchmark brought both of them much closer.

```
BenchmarkNetSMTP-8                 	    2000	    691668 ns/op	  128368 B/op	     338 allocs/op
BenchmarkEmersionGoSMTP-8          	    3000	    569822 ns/op	  122953 B/op	     388 allocs/op
```

Next, I need to test your https://github.com/emersion/go-message in place of my mimestream. However, I'm really curious as to why you basically forked the whole smtp package branching off to extend everything ([including textproto](https://github.com/emersion/go-message/commit/604a7372299371ce25c2a4dff3cbb2675a65a514)) manually.

Seems like you will miss out on any future work/issues on the stdlib repo if you don't keep a constant eye on it...

**Comment by Xeoncross** [comment]
Here is a profile of the benchmark. 

    go test -bench=. --benchmem -cpuprofile profile.out
    go tool pprof profile.out

![profile001](https://user-images.githubusercontent.com/56460/58443136-c7246d80-80b5-11e9-9b25-366ea42eaa84.png)

**Comment by emersion** [comment]
>However, I'm really curious as to why you basically forked the whole smtp package branching off to extend everything (including textproto) manually.

For the smtp package: "The smtp package is frozen and is not accepting new features." (from the official docs)

For the mime/multipart and net/textproto package: these mutate the incoming message because the header fields are re-ordered (stored in a map). This breaks DKIM signatures in e-mail messages. These are not frozen so only the parts I really needed to change have been forked.

**Comment by Xeoncross** [comment]
Turns out I was doing a [reverse DNS lookup](https://github.com/Xeoncross/smtpd/blob/master/smtpd.go#L194) on every connection which was most of the remaining difference. At this point there isn't much difference left between the server implementations. (Though this package has a matching client) which I do not.

```
BenchmarkNetSMTP-8                 	    2000	    622986 ns/op	  137102 B/op	     328 allocs/op
BenchmarkEmersionGoSMTP-8          	    2000	    657470 ns/op	  124408 B/op	     390 allocs/op
```

**Comment by emersion** [comment]
Makes sense!

**Comment by Xeoncross** [comment]
I added [go-message](https://github.com/emersion/go-message/blob/master/mail/reader.go) to compare with [mimestream](https://github.com/Xeoncross/mimestream). Again, it's hard to see a clear winner. `go-message` introduces almost 100 extra allocs, but is 6% faster.

```
BenchmarkMySMTP (with mimestream)-8                        	    3000	    539898 ns/op	  136741 B/op	     328 allocs/op
BenchmarkEmersionGoSMTP (with go-message)-8                 	    2000	    608130 ns/op	  107644 B/op	     487 allocs/op
BenchmarkEmersionGoSMTP (with mimestream)-8   	    2000	    637666 ns/op	  124434 B/op	     391 allocs/op
```

**Comment by emersion** [comment]
>go-message introduces almost 100 extra allocs

That doesn't surprise me too much. We need to allocate more complicated structures to keep track of header fields, and we need to introduce an extra buffer copy when reading fields to make sure whitespace in header fields is preserved.

**Comment by Xeoncross** [comment]
Is it only because of DKIM that you are doing all this extra work with textproto and the extra structures, or are are their other reasons?

**Comment by emersion** [comment]
It's generally assumed that messages won't be mutated in any way. This is true for DKIM, but also for IMAP for instance (which assumes messages are immutable). The header fields ordering issue is probably the most important one, since all e-mail processing agents are supposed to add headers only at the top (breaking this makes it impossible to understand which agent added which header field).

**Comment by Xeoncross** [comment]
Either using a [io.ReadSeeker you can rewind](https://golang.org/src/net/http/fs.go#L199), [buffer.Peek()](https://golang.org/src/bufio/bufio.go?s=3366:3410#L119), or just reading from an io.Reader and then using io.Multireader(read, original) to combine it back into the stream.

**Comment by emersion** [comment]
If you need to do a `Header.Delete` operation (also needed when doing a `Header.Set` to remove any existing field), then you need to remove bytes from the header and it starts getting annoying.

**Comment by Xeoncross** [comment]
Why would you mutate the headers? (I've just never thought about changing them)

**Comment by emersion** [comment]
In general you can't change an existing one, but you can delete it add add a new one on top. This is used e.g. for `Authentication-Results` and other header fields that are supposed to appear only once.

**Comment by emersion** [comment]
Closing since this issue isn't actionable.
